src/Gateway.sol

```diff
@@ -113,226 +71,174 @@ contract Gateway is IGateway, IExecutor, IUpgradable, GatewayEIP712 {
+    function admin() external view returns (address) {
+        return ERC1967.getAdmin();
     }
 
-    // Initialize networks
-    function _updateNetworks(Network[] calldata networks) private {
-        for (uint256 i = 0; i < networks.length; i++) {
-            Network calldata network = networks[i];
-            NetworkInfo storage info = _networkInfo[network.id];
-            require(info.domainSeparator == bytes32(0), "network already initialized");
-            require(network.id != NETWORK_ID || network.gateway == address(this), "wrong gateway address");
-            info.domainSeparator = computeDomainSeparator(network.id, network.gateway);
-            info.gasLimit = 15_000_000; // Default to 15M gas
-            info.relativeGasPrice = UFloatMath.ONE;
-            info.baseFee = 0;
-        }
+    function setAdmin(address newAdmin) external {
+        require(msg.sender == this.admin(), "unauthorized");
+        ERC1967.setAdmin(newAdmin);
     }
 
-    function _registerKeys(TssKey[] memory keys) private {
-        // We don't perform any arithmetic operation, except iterate a loop
-        unchecked {
-            // Register or activate tss key (revoked keys keep the previous nonce)
-            for (uint256 i = 0; i < keys.length; i++) {
-                TssKey memory newKey = keys[i];
-
-                // Read shard from storage
-                bytes32 shardId = _tssKeyToShardId(newKey);
-                KeyInfo storage shard = _shards[shardId];
-                uint8 status = shard.status;
-                uint32 nonce = shard.nonce;
-
-                // Check if the shard is not active
-                require((status & SHARD_ACTIVE) == 0, "already active, cannot register again");
-
-                // Check y-parity
-                uint8 yParity = newKey.yParity;
-                require(yParity == (yParity & 1), "y parity bit must be 0 or 1, cannot register shard");
-
-                // If nonce is zero, it's a new shard.
-                // If the shard exists, the provided y-parity must match the original one
-                uint8 actualYParity = uint8(BranchlessMath.toUint((status & SHARD_Y_PARITY) > 0));
-                require(
-                    nonce == 0 || actualYParity == yParity,
-                    "the provided y-parity doesn't match the existing y-parity, cannot register shard"
-                );
-
-                // if is a new shard shard, set its initial nonce to 1
-                shard.nonce = uint32(BranchlessMath.ternaryU32(nonce == 0, 1, nonce));
-
-                // enable/disable the y-parity flag
-                status = BranchlessMath.ternaryU8(yParity > 0, status | SHARD_Y_PARITY, status & ~SHARD_Y_PARITY);
-
-                // enable SHARD_ACTIVE bitflag
-                status |= SHARD_ACTIVE;
-
-                // Save new status in the storage
-                shard.status = status;
-            }
+    function upgrade(address newImplementation) external payable {
```

Comment:
Don't understand what was the issue with the old `upgradeAndCall`, also very often upgrade the gateway also requires running a migration script.

Reply:
so you're saying initialize is a migration procedure? why not call it migrate then?

Reply:
In any case when we discussed the gmp api you didn't mention it

Reply:
Because `upgradeAndCall` can be atomic, which is useful once the contract can be in a bad state if we do it in two different calls. (also if the migration reverts, the upgrade also reverts).

> In any case when we discussed the gmp api you didn't mention it

I didn't review every aspect of the new API, there are other aspects I think worth change in the API instead the gateway, I focused in the things that affects the gateway most, ex: GatewayOps. Also there's a few things I did which were optimizations (like using float instead numerator/denominator) which I didn't pointed out in my review because doesn't worth discussing it while the core wasn't ready.

Reply:
ok, we can revert this part. although `initialize` should be called `migrate` then its clear what its for.